### PR TITLE
Feature/integrate mmr srv

### PR DESCRIPTION
**Pull request by DaviRain-Su**
______

For contributor use:

- [ ] Added a changelog entry, using [`unclog`](https://github.com/informalsystems/unclog).
- [ ] If applicable: Unit tests written, added test to CI.
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Updated relevant documentation (`docs/`) and code comments.
- [ ] Re-reviewed `Files changed` in the Github PR explorer.

### CODE REVIEW

File: relayer/src/supervisor.rs [hunk 1]
```diff
@@ -543,6 +629,53 @@ fn init_subscriptions<Chain: ChainHandle>(
...
+    registry: &mut Registry<Chain>,
+) -> Result<Vec<(Chain, BeefySubscription)>, Error> {
+    let chains = &config.chains;
+
+    let mut beefy_subs = Vec::with_capacity(chains.len());
+
+    for chain_config in chains {
+        // check chain type,only substrate app chain can be subscribed beefy msg
+        let account_prefix = chain_config.account_prefix.clone();
+        match account_prefix.as_str() {
```
Comment (DaviRain-Su) [inline relayer/src/supervisor.rs:645]: Need Imovement

File: relayer/src/channel.rs [hunk 2]
```diff
@@ -1087,6 +1100,12 @@ impl<ChainA: ChainHandle, ChainB: ChainHandle> Channel<ChainA, ChainB> {
 
         // Build message(s) to update client on destination
         let mut msgs = self.build_update_client_on_dst(proofs.height())?;
+        // let tm = TrackedMsgs::new(msgs, "update client on destination for ChannelOpenConfirm");
```
Comment (DaviRain-Su) [inline relayer/src/channel.rs:1103]: Why notion this?

File: relayer/src/channel.rs [hunk 3]
```diff
@@ -987,6 +993,12 @@ impl<ChainA: ChainHandle, ChainB: ChainHandle> Channel<ChainA, ChainB> {
 
         // Build message(s) to update client on destination
         let mut msgs = self.build_update_client_on_dst(proofs.height())?;
+        // let tm = TrackedMsgs::new(msgs, "update client on destination for ChannelOpenAck");
```
Comment (DaviRain-Su) [inline relayer/src/channel.rs:996]: Why notion this?

File: relayer/src/channel.rs [hunk 4]
```diff
@@ -883,6 +882,12 @@ impl<ChainA: ChainHandle, ChainB: ChainHandle> Channel<ChainA, ChainB> {
 
         // Build message(s) to update client on destination
         let mut msgs = self.build_update_client_on_dst(proofs.height())?;
+        // let tm = TrackedMsgs::new(msgs, "update client on destination for ChannelOpenTry");
```
Comment (DaviRain-Su) [inline relayer/src/channel.rs:885]: Why notion this?

File: modules/src/clients/ics07_tendermint/client_def.rs [hunk 5]
```diff
@@ -81,8 +88,13 @@ impl ClientDef for TendermintClient {
                 None => None,
             };
 
-        let trusted_consensus_state =
-            downcast_consensus_state(ctx.consensus_state(&client_id, header.trusted_height)?)?;
+        // let trusted_consensus_state =
```
Comment (DaviRain-Su) [inline modules/src/clients/ics07_tendermint/client_def.rs:91]: need restore , dont's not need change this.

File: modules/src/clients/ics07_tendermint/client_def.rs [hunk 6]
```diff
@@ -186,10 +198,18 @@ impl ClientDef for TendermintClient {
             }
         }
 
-        Ok((
-            client_state.with_header(header.clone()),
-            ConsensusState::from(header),
-        ))
+        //TODO: remove this after test
```
Comment (DaviRain-Su) [inline modules/src/clients/ics07_tendermint/client_def.rs:201]: need restore , dont's not need change this.

File: modules/src/clients/ics07_tendermint/client_def.rs [hunk 7]
```diff
@@ -203,17 +223,22 @@ impl ClientDef for TendermintClient {
...
             epoch: consensus_height.revision_number,
             height: consensus_height.revision_height,
         };
+        tracing::trace!(target:"ibc-rs","[ics07_tendermint::client_def] verify_client_consensus_state path : {:?}",path);
+
         let value = expected_consensus_state
             .encode_vec()
             .map_err(Ics02Error::invalid_any_consensus_state)?;
-        verify_membership(client_state, prefix, proof, root, path, value)
+        // verify_membership(client_state, prefix, proof, root, path, value)
```
Comment (DaviRain-Su) [inline modules/src/clients/ics07_tendermint/client_def.rs:240]: need restore , dont's not need change this.

File: modules/src/core/ics04_channel/msgs/acknowledgement_response.rs [hunk 8]
```diff
@@ -36,6 +36,12 @@ impl Acknowledgement {
             acknowledgement::Response::Error(_e) => Ok(false),
         }
     }
+    pub fn acknowledgement(&self) -> Result<Vec<u8>, Error> {
```
Comment (DaviRain-Su) [inline modules/src/core/ics04_channel/msgs/acknowledgement_response.rs:39]: maybe dont's need this

File: modules/src/core/ics23_commitment/merkle.rs [hunk 9]
```diff
@@ -91,20 +98,35 @@ impl MerkleProof {
...
+
+                    return Err(Error::invalid_merkle_proof());
+                }
             }
         }
 
-        if root.hash != subroot {
-            return Err(Error::verification_failure());
-        }
+        // if root.hash != subroot {
```
Comment (DaviRain-Su) [inline modules/src/core/ics23_commitment/merkle.rs:124]: need restore , dont's not need change this.

File: relayer/src/chain/cosmos.rs [hunk 10]
```diff
@@ -1336,15 +1352,17 @@ impl ChainEndpoint for CosmosSdkChain {
...
         let consensus_state = AnyConsensusState::decode_vec(&res.value).map_err(Error::decode)?;
+        tracing::trace!(target:"ibc-rs","in cosmos: [proven_client_consensus] consensus_state: {:?}", consensus_state,);
 
-        if !matches!(consensus_state, AnyConsensusState::Tendermint(_)) {
-            return Err(Error::consensus_state_type_mismatch(
-                ClientType::Tendermint,
-                consensus_state.client_type(),
-            ));
-        }
+        // if !matches!(consensus_state, AnyConsensusState::Tendermint(_)) {
```
Comment (DaviRain-Su) [inline relayer/src/chain/cosmos.rs:1357]: need restore , dont's not need change this.

File: relayer/src/chain/substrate.rs [hunk 11]
```diff
@@ -624,8 +626,36 @@ impl ChainEndpoint for SubstrateChain {
         Ok((event_receiver, monitor_tx))
     }
 
+    fn init_beefy_monitor(
```
Comment (DaviRain-Su) [inline relayer/src/chain/substrate.rs:629]: maybe init beefy monitor can move init_event, so we not need have this function in ChainEndpoint trait.

File: relayer/src/chain/substrate.rs [hunk 12]
```diff
@@ -1577,23 +1945,60 @@ impl ChainEndpoint for SubstrateChain {
...
     fn query_host_consensus_state(&self, height: ICSHeight) -> Result<Self::ConsensusState, Error> {
-        tracing::trace!("in substrate: [query_host_consensus_state]");
+        tracing::trace!(target:"ibc-rs","in substrate: [query_host_consensus_state]");
 
         Ok(AnyConsensusState::Grandpa(GPConsensusState::default()))
     }
 }
 
+/// send Update client state request
+pub async fn send_update_state_request(
```
Comment (DaviRain-Su) [inline relayer/src/chain/substrate.rs:1961]: need move octopusxt

File: relayer/src/foreign_client.rs [hunk 13]
```diff
@@ -911,44 +961,44 @@ impl<DstChain: ChainHandle, SrcChain: ChainHandle> ForeignClient<DstChain, SrcCh
 
         */
 
-        let client_state = match client_state {
```
Comment (DaviRain-Su) [inline relayer/src/foreign_client.rs:914]: can remove

File: relayer/src/object.rs [hunk 14]
```diff
@@ -142,6 +166,8 @@ impl Packet {
 pub enum Object {
     /// See [`Client`].
     Client(Client),
+    /// See [`Beefy`]
```
Comment (DaviRain-Su) [inline relayer/src/object.rs:169]: means beefy don't at there.

File: modules/src/clients/ics07_tendermint/client_def.rs [hunk 15]
```diff
@@ -81,8 +88,13 @@ impl ClientDef for TendermintClient {
...
             };
 
-        let trusted_consensus_state =
-            downcast_consensus_state(ctx.consensus_state(&client_id, header.trusted_height)?)?;
+        // let trusted_consensus_state =
+        //     downcast_consensus_state(ctx.consensus_state(&client_id, header.trusted_height)?)?;
+        // FIXME: mismatch between client and arguments types
+        let pre_cs = ctx
+            .consensus_state(&client_id, header.trusted_height)
+            .unwrap();
```
Comment (DaviRain-Su) [inline modules/src/clients/ics07_tendermint/client_def.rs:96]: need roolback

File: modules/src/clients/ics07_tendermint/client_def.rs [hunk 16]
```diff
@@ -226,13 +251,24 @@ impl ClientDef for TendermintClient {
...
         client_state.verify_height(height)?;
 
         let path = ConnectionsPath(connection_id.clone());
+        tracing::trace!(target:"ibc-rs","[ics07_tendermint::client_def] verify_connection_state path : {:?}",path);
+
         let value = expected_connection_end
             .encode_vec()
             .map_err(Ics02Error::invalid_connection_end)?;
         verify_membership(client_state, prefix, proof, root, path, value)
+        // Ok(())
```
Comment (DaviRain-Su) [inline modules/src/clients/ics07_tendermint/client_def.rs:271]: need delete

File: modules/src/clients/ics07_tendermint/client_def.rs [hunk 17]
```diff
@@ -246,13 +282,18 @@ impl ClientDef for TendermintClient {
...
         client_state.verify_height(height)?;
 
         let path = ChannelEndsPath(port_id.clone(), *channel_id);
+        tracing::trace!(target:"ibc-rs","[ics07_tendermint::client_def] verify_channel_state path : {:?}",path);
+
         let value = expected_channel_end
             .encode_vec()
             .map_err(Ics02Error::invalid_channel_end)?;
         verify_membership(client_state, prefix, proof, root, path, value)
+        // Ok(())
```
Comment (DaviRain-Su) [inline modules/src/clients/ics07_tendermint/client_def.rs:296]: need delete

File: modules/src/clients/ics07_tendermint/client_def.rs [hunk 18]
```diff
@@ -265,13 +306,18 @@ impl ClientDef for TendermintClient {
...
         client_state.verify_height(height)?;
 
         let path = ClientStatePath(client_id.clone());
+        tracing::trace!(target:"ibc-rs","[ics07_tendermint::client_def] verify_client_full_state path : {:?}",path);
+
         let value = expected_client_state
             .encode_vec()
             .map_err(Ics02Error::invalid_any_client_state)?;
         verify_membership(client_state, prefix, proof, root, path, value)
+        // Ok(())
```
Comment (DaviRain-Su) [inline modules/src/clients/ics07_tendermint/client_def.rs:320]: need delete

File: modules/src/clients/ics07_tendermint/client_def.rs [hunk 19]
```diff
@@ -429,6 +496,7 @@ fn verify_membership(
             0,
         )
         .map_err(|e| Ics02Error::tendermint(Error::ics23_error(e)))
+    // Ok(())
```
Comment (DaviRain-Su) [inline modules/src/clients/ics07_tendermint/client_def.rs:499]: need delete

**Review by boern — APPROVED** [review]

File: modules/src/core/ics02_client/handler/update_client.rs [hunk 20]
```diff
@@ -94,11 +109,15 @@ pub fn process(
...
         height: ctx.host_height(),
         client_id,
-        ..Default::default()
+        client_type,
+        consensus_height: header_height,
     };
     output.emit(IbcEvent::UpdateClient(event_attributes.into()));
 
-    Ok(output.with_result(result))
+    let output = output.with_result(result);
```
Comment (DaviRain-Su) [inline modules/src/core/ics02_client/handler/update_client.rs:117]: roolback

File: modules/src/core/ics23_commitment/merkle.rs [hunk 21]
```diff
@@ -91,20 +98,35 @@ impl MerkleProof {
...
+
+                _ => {
+                    tracing::trace!(target:"ibc-rs","[ics23_commitment]  verify_membership not exist proof : {:?}",proof.proof);
+
+                    return Err(Error::invalid_merkle_proof());
+                }
             }
         }
 
-        if root.hash != subroot {
```
Comment (DaviRain-Su) [inline modules/src/core/ics23_commitment/merkle.rs:105]: roolback

**Pull request merged**